keps/sig-apps/3715-indexed-job-with-unset-completions/README.md

```diff
@@ -0,0 +1,887 @@
+issue in [kubernetes/enhancements] referencing this KEP and targeting a release
+milestone **before the [Enhancement Freeze](https://git.k8s.io/sig-release/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core
+Kubernetes—i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [x] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [x] (R) Design details are appropriately documented
+- [x] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [x] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Currently Jobs are validated to have `spec.completions` set when `spec.completionMode`=`Indexed`. This KEP proposes to relax this validation and allow 
+Indexed jobs to be created with zero `sepc.completions`. 
+
+When `spec.completions` is not set for an `Indexed` Job, the behavior is summarized as follows:
+- The success semantics is the same as `NonIndexed`. Specifically, the job is considered successful if any pod completes successfully. When this condition is achieved, running pods will be allowed to complete on their own and will not be explicitly deleted.
```

Comment:
> When this condition is achieved, running pods will be allowed to complete on their own and will not be explicitly deleted.

As we discussed previously, this seems a bit problematic in semantics, a few thoughts:  

1. Job is considered successful, but pods continue to run. Can it be confusing? The concept of "finished" becomes blurry. For example how would activeDeadlineSeconds and ttlAfterFinished be defined in this case?

2. The fate of pod retries are decided by other pods succeeding, which I find non-local and inconsistent. Before any pod succeeds, the pods keep retrying, but after some other pod succeeds, they don't.

3. What happens if parallelism is increased after one pod already succeeded? Is it ignored although existing pods are allowed to complete?

Please consider an alternative of allowing scaling completions together with parallelism. There would be two clean ways to successfully complete a Job with workers:
1. Finishing pods with success according to 'completions'.
2. Scaling completions to 0, which would stop all pods.

Reply:
Thanks Qelo, great points.

> As we discussed previously, this seems a bit problematic in semantics, a few thoughts:

This is the existing behavior of `completions=nil` for `NonIndexed` Job, so we are kinda stuck with it. I think changing this semantic for `Indexed` job would be a bigger inconsistency. 

> Job is considered successful, but pods continue to run. Can it be confusing? The concept of "finished" becomes blurry. For example how would activeDeadlineSeconds and ttlAfterFinished be defined in this case?

For completions=nil, the job is considered *successful* if any pod exits with success, the job is complete when there is at least one pod exited with success and all active pods existed. So activeDeadlineSeconds and ttlAfterFinished are tested against the completion timestamp.


> The fate of pod retries are decided by other pods succeeding, which I find non-local and inconsistent. Before any pod succeeds, the pods keep retrying, but after some other pod succeeds, they don't.

True, I think this is a bit strange. I think the original goal of this behavior was to support the work queue use case: https://kubernetes.io/docs/tasks/job/fine-parallel-processing-work-queue/

> What happens if parallelism is increased after one pod already succeeded? Is it ignored although existing pods are allowed to complete?

Yes.

The points above are valid, but lets also consider the use cases that we aim to target and see if the existing semantics handle it. The case we have in mind is the workers in a manager-worker batch workload patterns. Workers don't exit on their own, they are either scaled up or down via the parallelism parameter. In this scenario, success is really scaling parallelism down to 0.

> Please consider an alternative of allowing scaling completions together with parallelism.
> Finishing pods with success according to 'completions'.
> Scaling completions to 0, which would stop all pods.

This can be done in addition to this proposal, I don't see it as an alternative. But here are you thoughts on your proposal:

1. `completions` is an immutable field, we could make it mutable for the two cases mentioned above (can only change with parallelism and should be equal to parallelism), but this means that parallelism and completions should change together. We have plans to introduce scale subresource to Job, and I don't know if the scale subresource supports updating two parameters (I think you can point one replica parameter). 

An alternative is one we discussed offline: have a `successPolicy` parameter that allows users to modify the default success semantics. For `completions=nil`, we could have a policy that says `parallelism` number of pods must successfully exit to consider the job successful.

2. In general changing the completion semantics on the fly can be subject to race conditions, what if we declared the job successful and complete before the controller observed that the completion condition changed. But maybe this is solvable.

Reply:
> So activeDeadlineSeconds and ttlAfterFinished are tested against the completion timestamp.

I don't think that's the current implementation. If the Finished condition is not set yet, we just check against the startTime. If the condition is set, we simply don't sync the job anymore. The condition is only set once job tracking has finished counting all the pods.

Having completions immutable adds the benefit that we could inject the value into an environment variable, as some users have requested: https://github.com/kubernetes/kubernetes/issues/106009

Long term we probably want to have more flexibility around success. So probably starting with the same behavior that NonIndexed jobs have is a good default.

> In general changing the completion semantics on the fly can be subject to race conditions, what if we declared the job successful and complete before the controller observed that the completion condition changed. 

This would trigger conflict error when updating the Job status, given that the resourceVersion wouldn't match. One of the updates would happen first: changing the completions or adding the finished condition, but not both.

Reply:
> I don't think that's the current implementation. 

Partially right, I wasn't being literal, my point is that the finished condition is set when all the pods exit, not when the on the first pod to exit with success. btw, in the ttlAfterFinish case, we do test against the finished condition timestamp. I will lookup the exact semantics and explicitly document them in the KEP.

> This would trigger conflict error when updating the Job status, given that the resourceVersion wouldn't match. One of the updates would happen first: changing the completions or adding the finished condition, but not both.

Nice, so there is an inherit solution to this race.

Reply:
> So activeDeadlineSeconds and ttlAfterFinished are tested against the completion timestamp.

I'm looking here: https://github.com/kubernetes/kubernetes/blob/4802d7bb62c2623be8e4f940f6b5c1fcddd6c744/pkg/controller/job/job_controller.go#L802-L809
which clearly shows that we require at least 1 success and no running pods, which might be problematic in some edge cases. We might want to align the implementation to be closer to what @ahg-g mentioned, as in mark the job as completed (or maybe awaiting completion) and let the rest of pods to finish, that's not the case today.

Reply:
> We might want to align the implementation to be closer to what @ahg-g mentioned, as in mark the job as completed (or maybe awaiting completion) and let the rest of pods to finish, that's not the case today.

This would be a change of behavior that will surprise users. We need to have successPolicy and perhaps completionPolicy in the future to tune the default behavior.

Reply:
@soltysh I updated the KEP to describe the current behavior in more detail. PTAL.

Reply:
@ahg-g Could you expand the user story with details how the setup with workers finishes? Is the manager scaling down parallelism to 0 at the end, or workers complete with success, or maybe you want to support multiple ways?

Asking because it might help guide success semantics (first pod succeeding controversies etc.). For example, what if we said that Job with completions=nil can only succeed by setting parallelism=0 or by all pods succeeding? Would we lose something in terms of user story coverage?

And I guess some argument might be consistency with NonIndexed mode, but I would say it's more important to have simple clean semantics in IndexedMode than to be consistent with NonIndexed mode.

(I also considered "can only succeed by setting parallelism=0", which is a potential interpretation of completions=nil -> no number of completions will satisfy this job. But it's unclear what to do if all pods do succeed. )

Reply:
The driver/manager is the one that decides the success of the workload. That driver could either be index 0 or a completely different job. The latter is probably the best approach to setup such workloads, and the one we need to promote as a best practice because it is more flexible (allows to spec the manager differently in terms of resources, retries etc.). Once the driver exits, the workers should too, and their success/failure doesn't matter (e.g., the workers job gets deleted or scaled to zero once the manager decides that the workload is done)

So perhaps a success policy for completions=nil could be index 0 exiting successfully, and also the remaining pods are immediately removed.

> And I guess some argument might be consistency with NonIndexed mode, but I would say it's more important to have simple clean semantics in IndexedMode than to be consistent with NonIndexed mode.

I am not sure, @liggitt @soltysh what do you think? I am inclined to keep the default semantics exactly the same as `NonIndexed` and add a `spec.successPolicy` field to the Job API if we are to add new behavior.

Reply:
It would be reasonable for readers to assume spec.completions is non-nil if spec.completionMode=Indexed (since that was the explicitly documented requirement/relationship between those two fields)... I'm not sure we can/should start allowing setting that to `nil`

The implications for using parallelism for scaling and auto-inferring success criteria with completions set to nil in Indexed mode are also pretty non-obvious (at least to me)

As a way to evaluate potential designs here, I would recommend considering two aspects:
1. How would the documentation at https://kubernetes.io/docs/concepts/workloads/controllers/job/#parallel-jobs and https://kubernetes.io/docs/concepts/workloads/controllers/job/#completion-mode be updated/changed?
2. Could someone who read the current version of those topics and wrote code based on it have their code broken by the proposed changes?

Possible alternatives that come to mind:
* Keep requiring spec.completions but make it mutable along with spec.parallelism (and maybe only allow increasing spec.completions?)
* Have a new completion mode (like `IndexedWithScaling` or something) that uses indexing but expects the number of desired completions to scale

Reply:
>  I'm not sure we can/should start allowing setting that to nil

This is relaxing a validation, nothing will break for existing jobs if we start allowing that, and allowing it is not a security risk that would surprise users.

> The implications for using parallelism for scaling and auto-inferring success criteria with completions set to nil in Indexed mode are also pretty non-obvious (at least to me)

Nothing special about scaling, which part is not obvious? Regarding success, we are not proposing to change anything compared to the existing semantics for `NonIndexed`, a job is successful if at least one pod is successful and failures don't exceed backoffLimit by the time all pods completes.

> How would the documentation at
I took a stab at that here: https://github.com/kubernetes/kubernetes/issues/114862#issuecomment-1373945080

> Could someone who read the current version of those topics and wrote code based on it have their code broken by the proposed changes?

No. Here is one hypothetical example: https://github.com/kubernetes/kubernetes/issues/114862#issuecomment-1373905573

> Keep requiring spec.completions but make it mutable along with spec.parallelism (and maybe only allow increasing spec.completions?)

This is discussed as an alternative.

> Have a new completion mode (like IndexedWithScaling or something)

This was also discussed in https://github.com/kubernetes/kubernetes/issues/114862 and in the sig-apps meeting and had no support. I can add it as an alternative.

Reply:
>> Keep requiring spec.completions but make it mutable along with spec.parallelism (and maybe only allow increasing spec.completions?)
>
> This is discussed as an alternative.

It seems like there are two aspects involved here. The ability to scale an indexed job seems the primary goal. I can't tell if considering a job complete when the first pod succeeds is a goal as well, or if it's a side-effect of this proposed design we're trying to rationalize. Does it make sense that scaling up an indexed job means we want the first pod success to stop new pods from getting created? That seems like a weird interaction to me.

Letting spec.completions be mutable still seems clearer to me than making two different fields control the number of desired indexes and requiring a shift in the completion criteria.

* it keeps the concept of "how many completed indexes do I need" distinct from "how many pods do I want running at a time"
* I think it makes more sense to guard against races between updates to the completion target and an already completed job (changing completions on an already-completed job should fail)
* it allows scaling indexed jobs without changing how job completion criteria works

Reply:
Maybe we should consider leaving `parallelism` empty, and only in this case have mutable `completions`.

Although, that would imply changes in defaulting https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/batch/v1/defaults.go#L30

Could the defaulting depend on a new `completionMode`?

Reply:
> In general changing the completion semantics on the fly can be subject to race conditions, what if we declared the job successful and complete before the controller observed that the completion condition changed. But maybe this is solvable.

I don't think there would be these race conditions. We use Update when setting the status (and we should continue to do so), which is subject to collision errors if the spec changes at the same time.

Reply:
> It seems like there are two aspects involved here. The ability to scale an indexed job seems the primary goal. I can't tell if considering a job complete when the first pod succeeds is a goal as well, or if it's a side-effect of this proposed design we're trying to rationalize.

The goal is indeed to scale indexed job, and the proposal is trying to piggyback on the `completions=nil` semantics that already exist for `NonIndexed` mode. 

I think the success criteria is up for debate, and if we agree that it can diverge from the `NonIndexed` mode, then we can change it to: all remaining active pods successfully completing or scaled down.

> Letting spec.completions be mutable still seems clearer to me than making two different fields control the number of desired indexes and requiring a shift in the completion criteria.

I agree that making completions mutable is a strong alternative. But, `completions=nil` is not strictly new behavior for the Job API, and allowing it is a simpler change than making completions mutable.

> it keeps the concept of "how many completed indexes do I need" distinct from "how many pods do I want running at a time"

But for completions=parallelism, which is the case for the user story we are trying to enable, you will need to have both parameters scaled at the same time, this will not be possible to support via the scale subresource if we are to add it to  Job.

Moreover, the idea of scaling is not really related to how many pods I want completed, it is more about how many pods I want active, which is parallelism; that is why StatefulSet was considered for some of these workloads.

> it allows scaling indexed jobs without changing how job completion criteria works

I think saying that when completions=nil we fall back to parallelism is fair, at least this is what I would expect.

> it avoids exposing API clients expecting completions to always be non-nil for indexed jobs (based on our docs and previous behavior) to data they don't expect

This is also a problem for mutable completions, currently we say it doesn't change, and so clients may read it once and cache and not taking into account that it could now change.

> it seems like this proposal is going to encounter race conditions between scaling up the desired count and the already-existing pods completing regardless...

I don't think it does if we maintain the success semantics of `spec.completions=nil` for `NonIndexed` mode.

Reply:
Thanks all for the feedback, we had an offline discussion with Jordan, Greg and other folks during an api review meeting. 

We reached a conclusion that `spec.completions=nil` is not the ideal solution for the following reasons: 
1) it changes how completion criteria for Indexed job works; 
2) not safe for api clients that already assumed completion would always be non nil for Indexed Job 
3) the success semantics are not ideal and hard to reason about if we maintain the existing ones we have for `NonIndexed`
4) Adding scale subresource to job may never materialize (https://github.com/kubernetes/kubernetes/pull/58468#issuecomment-360794229)

To address those issues and other points mentioned above, I updated the proposal to introduce `ElasticIndexed` mode that allows mutating completions with parallelism in tandem.

The one uncertainty I have is whether to relax validation and make completions mutable on its own. I am hesitant  for two reasons: 1) there is no use case yet for it; 2) the idea of scaling is not really related to how many pods I want completed, it is more about how many pods I want active.

What do you folks think?

Reply:
Sounds good to me. Some extra thoughts:

1. What's the rationale for new mode vs allowing mutations in Indexed? (since we anyway avoided the problem of nil exception for api readers)

2. Will you allow scaling to parallelism=completions=0? It seems to make sense for the usecase of manager cleaning workers and I'd say the job should be marked as successful then. Btw. I still think it might be useful to list completion scenarios in user stories. In general for debugging and monitoring I think it's good if happy paths manifest itself as successes in statuses, not as pending or failed.

3. Regarding allowing completions to be mutable on its own (outside completions=parallelism), I'd lean towards allowing it -- if we don't see major semantics controversies with it. This would help avoid revisiting this later and another dilemma that readers might have assumed parallelism=completions. One thing I would keep validating, though, is parallelism <= completions (unless we don't we validate it today?). The case of parallelism > completions seems unclear for indexed.

Reply:
> To address those issues and other points mentioned above, I updated the proposal to introduce `ElasticIndexed` mode that allows mutating completions with parallelism in tandem.

I'd advise against mutating the completions after it's set (nil, or any other value), as mentioned just in this thread several times, the ability to change the rules for when a job reaches its completions while the controller operates is prone to races and should never be allowed. Parallelism is a different beast, since it controls how many pods are running in parallel so that can be freely modified always. Those were the guidelines we've picked when we initially wrong the jobs and based on everything I've experienced I haven't seen strong reasons to change those rules.

Reply:
> I updated the proposal to introduce ElasticIndexed mode that allows mutating completions with parallelism in tandem.

Do we need a new mode? Modifying completions+parallelism with mode=Indexed seems like a pretty clear indicator the job is elastic. I'm curious what controller code paths would require a distinct mode to trigger off of and couldn't just use the information about existing pods and completed indices to accomplish the same thing.

Reply:
> What's the rationale for new mode 
> Do we need a new mode? 

Why new mode: if we want to move forward with the idea that "elastic" means parallelism and completions are one, and should move together (i.e., parallelism can't be mutated on its own, same with completions, both should always be mutated together and with the same value), then a new mode is needed because this is not backward compatible with the semantics of parallelism in `Indexed`.

> Will you allow scaling to parallelism=completions=0?

I think completions=0 is success, but I will test it.

> Regarding allowing completions to be mutable on its own (outside completions=parallelism), I'd lean towards allowing it  if we don't see major semantics controversies with it

Yeah, the case where both move together is pretty clear to me semantically and easily maps to existing behaviors like StatefulSet, that is why I stated with a restricted scope. But so far I couldn't think of strange semantics evolving from allowing it to mutate on its own, but at the same time I can't think of a use case where this is needed. This is not enough to disallow it, hence the hesitation, so unless others can't think of any major consequences too, I would also allow it.

> I still think it might be useful to list completion scenarios in user stories

I will add that.

> One thing I would keep validating, though, is parallelism <= completions (unless we don't we validate it today?). The case of parallelism > completions seems unclear for indexed.

We don't validate that, but the controller handles this case by considering parallelism to be the max allowed active pods.

> I'd advise against mutating the completions after it's set 

Thanks @soltysh, I think the success semantics for those types of workloads will sometimes require all existing active pods to exit successfully, this is indeed much cleaner semantics. As to the race you mention, as others discussed, one of the two updates will get rejected anyways on update, and when it does, the job controller will requeue the job and process it again, so those races are already handled.

>  I'm curious what controller code paths would require a distinct mode to trigger off of and couldn't just use the information about existing pods and completed indices to accomplish the same thing.

I could find any, it seems that it will be handled just fine. But as I mentioned in other places, my only motivation was to have a distinct and easy to find API for a common use case we have in batch workloads, and limit the change to that case.

Reply:
@alculquicondor can you think of any changes needed to handle mutating completions unconditionally? If not, then I will update the KEP to propose that and remove the new enum.

Reply:
I could be persuaded to the option of Index with parallelism == completions with completions being mutable, without introducing new mode. That would be a reasonable middle ground, and probably will be best for users as well, although we'll need to document that sufficiently.

Reply:
Actually, it might not be too bad with indexed jobs, because we immediately update the completedIndexes field and don't care about tracking their UIDs in `uncountedTerminatedPods`. As a result, we wouldn't over count succeeded pods. But we might overcount failed pods if they fail at roughly the same time when completions is decreasing. But this sounds acceptable to me.

Reply:
Yeah, even in the tracking code we ignore pods with indexes outside of completions https://github.com/kubernetes/kubernetes/blob/b0ed87078e4042c12e456c8427dee0fb947a724c/pkg/controller/job/job_controller.go#L1002-LL1009
https://github.com/kubernetes/kubernetes/blob/b0ed87078e4042c12e456c8427dee0fb947a724c/pkg/controller/job/job_controller.go#L1044-L1048

Reply:
We should clarify in user docs that `.status.Failed` **won't** be decremented even if the failed pods are now outside the index range.

Reply:
> Regarding allowing completions to be mutable on its own.

From the job-controller perspective, it doesn't matter whether the fields are different.

The most important question is whether we want to ever support the `/scale` subresource. I don't see why this shouldn't be supported in the future, but then it requires that paralellism=completions to make sense. Can we do this in the strategy @liggitt? If anything, I think completions should dictate this, even though parallelism makes more syntactic sense.

Reply:
Right, that was my understanding as well with regards to success, since we count indexes, we should be fine. But the failed semantics indeed needs to be clarified to state that it covers the lifetime of the job, not only the active range.

Reply:
> Can we do this in the strategy @liggitt? If anything, I think completions should dictate this, even though parallelism makes more sense from a language perspective.

Are you thinking implicitly setting one to the other? I know we do that for defaulting on creation, but I think this would be surprising to clients in the update case.

Reply:
> I could be persuaded to the option of Index with parallelism == completions with completions being mutable, without introducing new mode. 

@soltysh what is your concern about allowing to mutate completions without this restriction? Mine was mostly about not having a case for it; but in terms of execution, it is not different than parallelism=completions, so I feel there is not strong reason to restrict it.

Reply:
I pushed a new update with the latest proposal: all mutating completions only if equal to parallelism for indexed jobs.

Reply:
We have one thing to decide on semantically:

In the case where an index successfully completes, and then the job is scaled down rendering the index
out of range, and then the job was scaled up again including back the index in the range, we have two options:

1) The index will be restarted, this is the behavior that we will get if we make no changes to the controller.
2) The index will be considered succeeded and not restarted. This requires a some change to the controller to keep those indexes in the `status.completedIndexes` across scale ups. Not a major issue to change.

Reply:
I vote for 1. If you are reducing the completions, you should be willing to forego the memory of higher indexes.